File: stealthwall/middleware/fastapi/stealthwall_fastapi/features.py
r"""STEALTHWALL — canonical feature extractor (Python side).

Implements docs/feature_extraction_spec.md v1 EXACTLY. The Node.js mirror
(middleware/express/src/features.js) must remain line-for-line equivalent;
tests/parity replays identical event sequences through both and fails CI on
any divergence.

Parity discipline (why the code looks the way it does):
- NO regex anywhere: Python `re` and JS RegExp disagree on unicode classes
  (e.g. `\d`); every normalization step is an explicit character loop that
  translates 1:1 to JavaScript.
- ASCII-only casing: `asciiLower` maps ONLY 0x41–0x5A, never locale-aware
  `str.lower()` / `toLowerCase()`.
- Half-up integer rounding via `floor(v * 10^d + 0.5)` — never Python's
  banker's `round()` nor JS `Math.round()` half-toward-zero quirks.
- All tuned numbers imported from config/defaults.py (single source).
"""

from __future__ import annotations

import math
from typing import Dict, List, Optional

import sys
from pathlib import Path

_CONFIG_DIR = Path(__file__).resolve().parents[3] / "config"
if str(_CONFIG_DIR.parent) not in sys.path:  # repo root on path for `config` pkg
    sys.path.insert(0, str(_CONFIG_DIR.parent))

from config.defaults import (  # noqa: E402
    FEATURE_ROUNDING_DECIMALS,
    FEATURE_SPEC_VERSION,
    PAYLOAD_SAMPLE_MAX_BYTES,
    WINDOW_MAX_EVENTS_PER_IP,
    WINDOW_SECONDS,
)

# Spec Section 4 patterns (ASCII lowercase literals; substring containment).
# These are SPEC-owned constants shared verbatim by the Node mirror — not
# operator-tuned numbers, so they do NOT live in config/defaults.py.
SIGNATURE_PATTERNS = [
    "..%2f", "..\\", "../", "<script", "javascript:",
    "onerror=", "onload=", "union select", "or 1=1", "' or '",
    "--", "; drop table", "../etc/passwd", "%00", "${jndi:", "${",
    "../../", "%27", "%20or%20", "<img", "eval(", "exec(",
    "system(", "base64_decode(", "information_schema", "waitfor delay",
    "benchmark(", "load_file(", "into outfile", "@@version",
]

FEATURE_KEYS = [
    "request_rate",
    "unique_path_ratio",
    "path_entropy",
    "notfound_ratio",
    "auth_failure_ratio",
    "avg_payload_entropy",
    "signature_score",
    "timing_variance",
    "header_anomaly_score",
    "method_post_ratio",
    "avg_path_depth",
    "digit_ratio_in_path",
    "user_agent_entropy",
    "window_utilization",
]

EXPECTED_HEADERS = ["host", "user-agent", "accept", "connection"]
SUSPICIOUS_HEADERS = [
    "x-original-url",
    "x-rewrite-url",
    "proxy-authorization",
    "x-custom-forwarded",
]


def ascii_lower(s: str) -> str:
    """Lowercase ONLY ASCII A-Z (0x41-0x5A). Never use s.lower()."""
    out = []
    for ch in s:
        code = ord(ch)
        out.append(chr(code + 32) if 65 <= code <= 90 else ch)
    return "".join(out)


def round_to(value: float) -> float:
    """Spec 3.5: floor(v * 10^d + 0.5) / 10^d — identical in JS."""
    scaled = value * (10 ** FEATURE_ROUNDING_DECIMALS)
    return math.floor(scaled + 0.5) / (10 ** FEATURE_ROUNDING_DECIMALS)


def normalize_path(path: str) -> str:
    """Spec 3.1 — explicit loops, no regex, ASCII-only casing."""
    qpos = path.find("?")
    if qpos != -1:
        path = path[:qpos]
    out = []
    prev_digit = False
    prev_slash = False
    for ch in path:
        code = ord(ch)
        is_digit = 48 <= code <= 57
        is_slash = ch == "/"
        if is_digit:
            if not prev_digit:
                out.append("N")
            prev_digit = True
            continue
        prev_digit = False
        if is_slash:
            if not prev_slash:
                out.append("/")
            prev_slash = True
            continue
        prev_slash = False
        if 65 <= code <= 90:
            out.append(chr(code + 32))
        else:
            out.append(ch)
    result = "".join(out)
    if len(result) > 1 and result.endswith("/"):
        result = result[:-1]
    return result


def segment_count(normalized_path: str) -> int:
    count = 0
    for part in normalized_path.split("/"):
        if len(part) > 0:
            count += 1
    return count


def truncate_payload_utf8(payload: str) -> str:
    """Truncate at UTF-8 BYTE boundary (spec Section 1) without splitting
    a multibyte sequence — decode with error replacement, identically in JS
    via TextDecoder(fatal=false)."""
    raw = payload.encode("utf-8")[:PAYLOAD_SAMPLE_MAX_BYTES]
    return raw.decode("utf-8", errors="replace")


def shannon_entropy(items: List[str]) -> float:
    """Spec 3.2, bits."""
    n = len(items)
    if n <= 1:
        return 0.0
    counts: Dict[str, int] = {}
    for it in items:
        counts[it] = counts.get(it, 0) + 1
    h = 0.0
    for c in counts.values():
        p = c / n
        h -= p * math.log2(p)
    return h


def byte_entropy(payload: str) -> float:
    """Spec 3.3, bits/byte over UTF-8 bytes of the truncated payload."""
    data = payload.encode("utf-8")
    n = len(data)
    if n == 0:
        return 0.0
    freq = [0] * 256
    for b in data:
        freq[b] += 1
    h = 0.0
    for c in freq:
        if c > 0:
            p = c / n
            h -= p * math.log2(p)
    return h


def population_variance(xs: List[float]) -> float:
    """Spec 3.4."""
    n = len(xs)
    if n < 2:
        return 0.0
    mean = sum(xs) / n
    acc = 0.0
    for x in xs:
        diff = x - mean
        acc += diff * diff
    return acc / n


def matches_signature(event_path: str, event_payload: str) -> bool:
    """Spec Section 4: plain substring containment on ASCII-lowercased
    RAW path+payload concatenation. No regex engines involved."""
    haystack = ascii_lower(event_path) + ascii_lower(event_payload)
    for pat in SIGNATURE_PATTERNS:
        if pat in haystack:
            return True
    return False


def header_anomaly_score_event(headers: Dict[str, str]) -> float:
    """Spec Section 5."""
    for name in SUSPICIOUS_HEADERS:
        if name in headers:
            return 1.0
    missing = 0
    for name in EXPECTED_HEADERS:
        if name not in headers:
            missing += 1
    return missing / 4


def _window(events: List[dict]) -> List[dict]:
    """Spec Section 2: sliding window ending at latest event ts."""
    if not events:
        return []
    ordered = sorted(events, key=lambda e: e["ts"])
    latest_ts = ordered[-1]["ts"]
    boundary = latest_ts - WINDOW_SECONDS
    kept = [e for e in ordered if e["ts"] >= boundary]
    if len(kept) > WINDOW_MAX_EVENTS_PER_IP:
        kept = kept[-WINDOW_MAX_EVENTS_PER_IP:]
    return kept


def extract_features(events: List[dict]) -> Optional[List[float]]:
    """Return the 14-float vector (spec Section 6 order) or None when the
    window is empty. Every returned value passes through roundTo()."""
    window = _window(events)
    n_events = len(window)
    if n_events == 0:
        return None

    normalized_paths: List[str] = []
    uas: List[str] = []
    entropies: List[float] = []
    notfound = 0
    auth_failures = 0
    sig_matches = 0
    anomalies: List[float] = []
    post_count = 0
    depths: List[int] = []
    total_chars = 0
    digit_chars = 0

    truncated: List[str] = []
    for e in window:
        tp = truncate_payload_utf8(e.get("payload", "") or "")
        truncated.append(tp)
        raw_path = e.get("path", "") or ""
        qpos_raw = raw_path.find("?")
        queryless = raw_path[:qpos_raw] if qpos_raw != -1 else raw_path
        for ch in queryless:
            code = ord(ch)
            if 48 <= code <= 57:  # digit ratio measured on RAW path (spec 6)
                digit_chars += 1
            total_chars += 1
        np_path = normalize_path(raw_path)
        normalized_paths.append(np_path)
        uas.append(e.get("user_agent", "") or "")
        entropies.append(byte_entropy(tp))
        if e.get("status") == 404:
            notfound += 1
        if e.get("is_auth_failure"):
            auth_failures += 1
        if matches_signature(e.get("path", "") or "", tp):
            sig_matches += 1
        anomalies.append(header_anomaly_score_event(e.get("headers", {}) or {}))
        if (e.get("method", "") or "") == "POST":
            post_count += 1
        depths.append(segment_count(np_path))

    # timing gaps over ts-sorted window
    gaps: List[float] = []
    for i in range(1, len(window)):
        gaps.append(window[i]["ts"] - window[i - 1]["ts"])

    denom = max(1, n_events)

    vec = [
        n_events / WINDOW_SECONDS,
        len(set(normalized_paths)) / denom,
        shannon_entropy(normalized_paths) / math.log2(max(2, n_events)),
        notfound / denom,
        auth_failures / denom,
        (sum(entropies) / denom) / 8.0,
        sig_matches / denom,
        population_variance(gaps),
        sum(anomalies) / denom,
        post_count / denom,
        min(1.0, (sum(depths) / denom) / 10.0),
        (digit_chars / total_chars) if total_chars > 0 else 0.0,
        shannon_entropy(uas) / math.log2(max(2, len(set(uas)))),
        min(1.0, n_events / WINDOW_MAX_EVENTS_PER_IP),
    ]
    return [round_to(v) for v in vec]


def feature_vector_envelope(vector: List[float]) -> dict:
    """Serialization envelope (spec Section 6): fixed-order array tagged
    with the spec version."""
    return {
        "feature_spec_version": FEATURE_SPEC_VERSION,
        "keys": FEATURE_KEYS,
        "vector": vector,
    }

File: stealthwall/middleware/express/src/features.js
/**
 * STEALTHWALL — canonical feature extractor (Node side).
 *
 * LINE-FOR-LINE MIRROR of middleware/fastapi/stealthwall_fastapi/features.py
 * against docs/feature_extraction_spec.md v1. tests/parity replays identical
 * event sequences through both implementations and fails CI on divergence.
 *
 * Parity discipline mirrors the Python docstring:
 *  - NO regex anywhere (JS RegExp vs Python re disagree on unicode classes);
 *    every normalization step is an explicit character loop identical to the
 *    Python one.
 *  - ASCII-only casing (never locale-aware toLowerCase()).
 *  - Half-up integer rounding via Math.floor(v * 10^d + 0.5) — never
 *    Math.round()'s half-toward-infinity on negatives asymmetry (values are
 *    non-negative here, but the formula matches Python exactly anyway).
 *  - All tuned numbers loaded from config.json (generated mirror of
 *    config/defaults.py via config/sync_config.py; parity-tested).
 */

'use strict';

const fs = require('fs');
const path = require('path');

// -- config bridge (single source of truth lives in Python defaults.py) -----
function loadConfig() {
  const candidates = [
    path.join(__dirname, '..', 'config.json'),
    path.join(process.cwd(), 'middleware', 'express', 'config.json'),
  ];
  for (const p of candidates) {
    if (fs.existsSync(p)) return JSON.parse(fs.readFileSync(p, 'utf8'));
  }
  throw new Error('config.json not found; run python3 config/sync_config.py');
}

const CONFIG = loadConfig();

// -- spec-owned constants (NOT operator-tuned; mirrored verbatim) ----------
const SIGNATURE_PATTERNS = [
  '..%2f', '..\\', '../', '<script', 'javascript:',
  'onerror=', 'onload=', 'union select', 'or 1=1', "' or '",
  '--', '; drop table', '../etc/passwd', '%00', '${jndi:', '${',
  '../../', '%27', '%20or%20', '<img', 'eval(', 'exec(',
  'system(', 'base64_decode(', 'information_schema', 'waitfor delay',
  'benchmark(', 'load_file(', 'into outfile', '@@version',
];

const EXPECTED_HEADERS = ['host', 'user-agent', 'accept', 'connection'];
const SUSPICIOUS_HEADERS = [
  'x-original-url', 'x-rewrite-url', 'proxy-authorization',
  'x-custom-forwarded',
];

const FEATURE_KEYS = [
  'request_rate', 'unique_path_ratio', 'path_entropy', 'notfound_ratio',
  'auth_failure_ratio', 'avg_payload_entropy', 'signature_score',
  'timing_variance', 'header_anomaly_score', 'method_post_ratio',
  'avg_path_depth', 'digit_ratio_in_path', 'user_agent_entropy',
  'window_utilization',
];

// -- spec Section 3 primitives ----------------------------------------------

function asciiLower(s) {
  // Lowercase ONLY A-Z (0x41-0x5A). Never use s.toLowerCase().
  let out = '';
  for (let i = 0; i < s.length; i++) {
    const code = s.charCodeAt(i);
    out += code >= 65 && code <= 90 ? String.fromCharCode(code + 32) : s[i];
  }
  return out;
}

function roundTo(value) {
  // Spec 3.5: floor(v * 10^d + 0.5) / 10^d — identical in Python.
  const d = CONFIG.FEATURE_ROUNDING_DECIMALS;
  const scaled = value * Math.pow(10, d);
  return Math.floor(scaled + 0.5) / Math.pow(10, d);
}

function normalizePath(p) {
  // Spec 3.1 — explicit loops, no regex, ASCII-only casing.
  const qpos = p.indexOf('?');
  if (qpos !== -1) p = p.slice(0, qpos);
  let out = '';
  let prevDigit = false;
  let prevSlash = false;
  for (let i = 0; i < p.length; i++) {
    const ch = p[i];
    const code = p.charCodeAt(i);
    const isDigit = code >= 48 && code <= 57;
    if (isDigit) {
      if (!prevDigit) out += 'N';
      prevDigit = true;
      continue;
    }
    prevDigit = false;
    if (ch === '/') {
      if (!prevSlash) out += '/';
      prevSlash = true;
      continue;
    }
    prevSlash = false;
    if (code >= 65 && code <= 90) out += String.fromCharCode(code + 32);
    else out += ch;
  }
  if (out.length > 1 && out.endsWith('/')) out = out.slice(0, -1);
  return out;
}

function segmentCount(normalizedPath) {
  let count = 0;
  for (const part of normalizedPath.split('/')) {
    if (part.length > 0) count++;
  }
  return count;
}

function truncatePayloadUtf8(payload) {
  // Truncate at UTF-8 BYTE boundary without splitting multibyte sequences;
  // Buffer.toString('utf8') replaces invalid bytes with U+FFFD exactly like
  // Python's errors='replace' decoder.
  const raw = Buffer.from(payload, 'utf8');
  const sliced = raw.slice(0, CONFIG.PAYLOAD_SAMPLE_MAX_BYTES);
  return sliced.toString('utf8');
}

function shannonEntropy(items) {
  // Spec 3.2, bits.
  const n = items.length;
  if (n <= 1) return 0.0;
  const counts = new Map();
  for (const it of items) counts.set(it, (counts.get(it) || 0) + 1);
  let h = 0.0;
  for (const c of counts.values()) {
    const p = c / n;
    h -= p * Math.log2(p);
  }
  return h;
}

function byteEntropy(payload) {
  // Spec 3.3, bits/byte over UTF-8 bytes of the truncated payload.
  const data = Buffer.from(payload, 'utf8');
  const n = data.length;
  if (n === 0) return 0.0;
  const freq = new Array(256).fill(0);
  for (let i = 0; i < n; i++) freq[data[i]]++;
  let h = 0.0;
  for (const c of freq) {
    if (c > 0) {
      const p = c / n;
      h -= p * Math.log2(p);
    }
  }
  return h;
}

function populationVariance(xs) {
  // Spec 3.4.
  const n = xs.length;
  if (n < 2) return 0.0;
  let mean = 0.0;
  for (const x of xs) mean += x;
  mean /= n;
  let acc = 0.0;
  for (const x of xs) {
    const diff = x - mean;
    acc += diff * diff;
  }
  return acc / n;
}

function matchesSignature(eventPath, eventPayload) {
  // Spec Section 4: plain substring containment on ASCII-lowercased RAW
  // path+payload concatenation. No regex engines involved.
  const haystack = asciiLower(eventPath) + asciiLower(eventPayload);
  for (const pat of SIGNATURE_PATTERNS) {
    if (haystack.includes(pat)) return true;
  }
  return false;
}

function headerAnomalyScoreEvent(headers) {
  // Spec Section 5.
  for (const name of SUSPICIOUS_HEADERS) {
    if (Object.prototype.hasOwnProperty.call(headers, name)) return 1.0;
  }
  let missing = 0;
  for (const name of EXPECTED_HEADERS) {
    if (!Object.prototype.hasOwnProperty.call(headers, name)) missing++;
  }
  return missing / 4;
}

function windowEvents(events) {
  // Spec Section 2: sliding window ending at latest event ts.
  if (!events || events.length === 0) return [];
  const ordered = [...events].sort((a, b) => a.ts - b.ts); // stable sort
  const latestTs = ordered[ordered.length - 1].ts;
  const boundary = latestTs - CONFIG.WINDOW_SECONDS;
  let kept = ordered.filter((e) => e.ts >= boundary);
  if (kept.length > CONFIG.WINDOW_MAX_EVENTS_PER_IP) {
    kept = kept.slice(kept.length - CONFIG.WINDOW_MAX_EVENTS_PER_IP);
  }
  return kept;
}

// -- Spec Section 6: the 14-feature vector -----------------------------------

function extractFeatures(events) {
  const win = windowEvents(events);
  const nEvents = win.length;
  if (nEvents === 0) return null;

  const normalizedPaths = [];
  const uas = [];
  const entropies = [];
  let notfound = 0;
  let authFailures = 0;
  let sigMatches = 0;
  const anomalies = [];
  let postCount = 0;
  const depths = [];
  let totalChars = 0;
  let digitChars = 0;

  for (const e of win) {
    const tp = truncatePayloadUtf8(e.payload || '');
    const rawPath = e.path || '';
    const qposRaw = rawPath.indexOf('?');
    const queryless = qposRaw !== -1 ? rawPath.slice(0, qposRaw) : rawPath;
    for (let i = 0; i < queryless.length; i++) {
      const code = queryless.charCodeAt(i);
      if (code >= 48 && code <= 57) digitChars++; // RAW path digits (spec 6)
      totalChars++;
    }
    const np = normalizePath(rawPath);
    normalizedPaths.push(np);
    uas.push(e.user_agent || '');
    entropies.push(byteEntropy(tp));
    if (e.status === 404) notfound++;
    if (e.is_auth_failure) authFailures++;
    if (matchesSignature(rawPath, tp)) sigMatches++;
    anomalies.push(headerAnomalyScoreEvent(e.headers || {}));
    if ((e.method || '') === 'POST') postCount++;
    depths.push(segmentCount(np));
  }

  // timing gaps over ts-sorted window
  const gaps = [];
  for (let i = 1; i < win.length; i++) gaps.push(win[i].ts - win[i - 1].ts);

  const denom = Math.max(1, nEvents);

  const vec = [
    nEvents / CONFIG.WINDOW_SECONDS,
    new Set(normalizedPaths).size / denom,
    shannonEntropy(normalizedPaths) / Math.log2(Math.max(2, nEvents)),
    notfound / denom,
    authFailures / denom,
    entropies.reduce((a, b) => a + b, 0.0) / denom / 8.0,
    sigMatches / denom,
    populationVariance(gaps),
    anomalies.reduce((a, b) => a + b, 0.0) / denom,
    postCount / denom,
    Math.min(1.0, depths.reduce((a, b) => a + b, 0.0) / denom / 10.0),
    totalChars > 0 ? digitChars / totalChars : 0.0,
    shannonEntropy(uas) / Math.log2(Math.max(2, new Set(uas).size)),
    Math.min(1.0, nEvents / CONFIG.WINDOW_MAX_EVENTS_PER_IP),
  ];
  return vec.map(roundTo);
}

function featureVectorEnvelope(vector) {
  return {
    feature_spec_version: CONFIG.FEATURE_SPEC_VERSION,
    keys: FEATURE_KEYS,
    vector,
  };
}

module.exports = {
  asciiLower,
  roundTo,
  normalizePath,
  segmentCount,
  truncatePayloadUtf8,
  shannonEntropy,
  byteEntropy,
  populationVariance,
  matchesSignature,
  headerAnomalyScoreEvent,
  extractFeatures,
  featureVectorEnvelope,
  FEATURE_KEYS,
  SIGNATURE_PATTERNS,
  CONFIG,
};
